### ISSUE TITLE
[Bug]: Renderer Crash (SIGSEGV / SEGV_PKUERR) in 26.0.0, 26.1.0 and 26.2.0

### ISSUE THREAD

**Issue by threema-danilo** [electron/electron#39775]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

26.2.0

### What operating system are you using?

Other Linux

### Operating System Version

Arch Linux

### What arch are you using?

x64

### Last Known Working Electron version

25.8.0

### Expected Behavior

The renderer should not crash.

### Actual Behavior

An application I'm working on (which isn't open source yet) crashes on start when using Electron 26.0.0, 26.1.0 or 26.2.0.

The crash dump is always identical. The crashing thread crashed in the function `electron!partition_alloc::internal::PartitionBucket::MaintainActiveList()`.

Here are the top 4 calls in the stack trace:

```
Operating system: Linux
                  6.4.12 -arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 24 Aug 2023 00:38:14 +0000 x86_64
CPU: amd64
     family 25 model 33 stepping 0
     32 CPUs

Crash reason:  SIGSEGV / SEGV_PKUERR
Crash address: 0x2c0a30001048
Process uptime: 4 seconds

Thread 0 electron (crashed)
 0  electron!partition_alloc::internal::PartitionBucket::MaintainActiveList() [partition_bucket.cc : 1163 + 0x0]
     rax = 0x00002c0a30001040    rdx = 0x00002c0a30001040
     rcx = 0x0000000000000000    rbx = 0x00005629523020b0
     rsi = 0x0000000010600002    rdi = 0x00002c0a30008020
     rbp = 0x00007ffcda1f5c30    rsp = 0x00007ffcda1f5c20
      r8 = 0x0000000010600002     r9 = 0xf08f0100dc070000
     r10 = 0x0000000000000000    r11 = 0x0000000000000001
     r12 = 0x0000000000000000    r13 = 0x0000562952250008
     r14 = 0x0000000000000000    r15 = 0x00005629523020b0
     rip = 0x000056294dbd18c3
    Found by: given as instruction pointer in context
 1  electron!partition_alloc::PartitionRoot::PurgeMemory(int) [partition_root.cc : 1380 + 0x7]
     rbx = 0x0000000000000003    rbp = 0x00007ffcda1f60c0
     rsp = 0x00007ffcda1f5c40    r12 = 0x0000000000000000
     r13 = 0x0000562952250008    r14 = 0x0000562952303488
     r15 = 0x00005629523020b0    rip = 0x000056294dbd671c
    Found by: call frame info
 2  electron!partition_alloc::MemoryReclaimer::Reclaim(int) [memory_reclaimer.cc : 89 + 0x7]
     rbx = 0x0000562952250008    rbp = 0x00007ffcda1f60f0
     rsp = 0x00007ffcda1f60d0    r12 = 0x00000234002c34e0
     r13 = 0x0000562952250008    r14 = 0x0000000000000003
     r15 = 0x0000562952250018    rip = 0x000056294dbd01f4
    Found by: call frame info
 3  electron!base::allocator::(anonymous namespace)::RunMemoryReclaimer(scoped_refptr<base::SequencedTaskRunner>) [partition_alloc_support.cc : 227 + 0x7]
     rbx = 0x0000023400239740    rbp = 0x00007ffcda1f6160
     rsp = 0x00007ffcda1f6100    r12 = 0x0000000436d8d651
     r13 = 0x0000562952250008    r14 = 0xaaaaaaaaaaaaaaaa
     r15 = 0x00007ffcda1f6200    rip = 0x000056294daeb5d8
    Found by: call frame info
```

Here are two symbolicated crash dumps for 26.0.0 and 26.2.0:

- [crash-26.0.0.txt](https://github.com/electron/electron/files/12549635/crash-26.0.0.txt)
- [crash-26.2.0.txt](https://github.com/electron/electron/files/12549638/crash-26.2.0.txt)

With 25.8.0, everything works fine. Let me know if there's any additional information I can provide!

**Comment by indutny-signal** [comment]
Just to add a data point: all of our internal Linux users experienced the very same crash (with the same stack trace)

**Comment by deepak1556** [comment]
Does it crash when you launch with the following flag `--disable-features=PartitionAllocBackupRefPtr`

**Comment by deepak1556** [comment]
If the above does not crash, then please also check if the crash is avoided with `--enable-features=PartitionAllocBackupRefPtr:brp-mode/enabled` which will disable the memory reclaimer feature enabled via https://chromium-review.googlesource.com/c/chromium/src/+/4597058.

**Comment by ayumi-signal** [comment]
@deepak1556 Thank you for the suggestion of flags to try. I tried to launch electron using `--disable-features=PartitionAllocBackupRefPtr` and separately with `--enable-features=PartitionAllocBackupRefPtr:brp-mode/enabled` and in both cases electron crashes with the same error and location. The traces look the same as the trace from running without any flags. This is on Debian 12.

```
Operating system: Linux
                  6.1.0 -11-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.38-4 (2023-08-08) x86_64
CPU: amd64
     family 6 model 186 stepping 2
     20 CPUs

Crash reason:  SIGSEGV / SEGV_PKUERR
Crash address: 0x000002f1d0001048
Process uptime: 5 seconds

Thread 0 electron (crashed)
 0  electron!partition_alloc::internal::PartitionBucket::MaintainActiveList() [partition_bucket.c
c : 1163 + 0x0]
     rax = 0x000002f1d0001040    rdx = 0x000002f1d0001040
     rcx = 0x0000000000000000    rbx = 0x0000562fbb9e90b0
     rsi = 0x0000000010600002    rdi = 0x000002f1d0008020
     rbp = 0x00007ffe21d7dfa0    rsp = 0x00007ffe21d7df90
      r8 = 0x0000000010600002     r9 = 0xc00f0100c81e0000
     r10 = 0x0000000000000000    r11 = 0x0000000000000001
     r12 = 0x0000000000000000    r13 = 0x0000562fbb937008
     r14 = 0x0000000000000000    r15 = 0x0000562fbb9e90b0
     rip = 0x0000562fb72a7e03
    Found by: given as instruction pointer in context
 1  electron!partition_alloc::PartitionRoot::PurgeMemory(int) [partition_root.cc : 1380 + 0x7]
     rbx = 0x0000000000000003    rbp = 0x00007ffe21d7e430
     rsp = 0x00007ffe21d7dfb0    r12 = 0x0000000000000000
     r13 = 0x0000562fbb937008    r14 = 0x0000562fbb9ea488
     r15 = 0x0000562fbb9e90b0    rip = 0x0000562fb72acd4c
    Found by: call frame info
 2  electron!partition_alloc::MemoryReclaimer::Reclaim(int) [memory_reclaimer.cc : 89 + 0x7]
     rbx = 0x0000562fbb937008    rbp = 0x00007ffe21d7e460
     rsp = 0x00007ffe21d7e440    r12 = 0x00001adc002d1b00
     r13 = 0x0000562fbb937008    r14 = 0x0000000000000003
     r15 = 0x0000562fbb937018    rip = 0x0000562fb72a6734
    Found by: call frame info
 3  electron!base::allocator::(anonymous namespace)::RunMemoryReclaimer(scoped_refptr<base::Seque
ncedTaskRunner>) [partition_alloc_support.cc : 227 + 0x7]
     rbx = 0x00001adc00221950    rbp = 0x00007ffe21d7e4d0
     rsp = 0x00007ffe21d7e470    r12 = 0x0000004999c79119
     r13 = 0x0000562fbb937008    r14 = 0xaaaaaaaaaaaaaaaa
     r15 = 0x00007ffe21d7e500    rip = 0x0000562fb71c2058
    Found by: call frame info
```

**Comment by deepak1556** [comment]
@ayumi-signal to clarify, is the above crash from your application or from a sample electron app ? If it is from your application, add the feature flag to this callsite https://github.com/signalapp/Signal-Desktop/blob/4d2910d696fc22fd74dd36c3fa3749fb855e184a/app/main.ts#L1677

```
app.commandLine.appendSwitch('disable-features', 'HardwareMediaKeyHandling,PartitionAllocBackupRefPtr');
```

**Comment by deepak1556** [comment]
Thanks for confirming, I wanted to confirm if the BRP feature that was enabled by default on Linux https://chromium-review.googlesource.com/c/chromium/src/+/4478673 between 114 and 116 was related to the regression. But now that I look at the brp enabled process, it is only set to [kBrowserProcess](https://source.chromium.org/chromium/chromium/src/+/refs/tags/116.0.5791.1:base/allocator/partition_alloc_features.cc;l=135) on Linux and given the crash is in renderer per the issue title the feature would be unrelated.


> Regarding the other flag --enable-features=PartitionAllocBackupRefPtr:brp-mode/enabled I wasn't sure how to use it.

You can ignore this flag, it only affects the memory reclaimer code path associated with the BRP feature. The crash you are seeing is from https://source.chromium.org/chromium/chromium/src/+/refs/tags/116.0.5845.188:base/allocator/partition_alloc_support.cc;l=1429-1430

I would recommend the following as next steps:

* Backport https://chromium-review.googlesource.com/c/chromium/src/+/4665034 which isolates the memory reclaimer into a separate feature and you should be able to disable the feature as a temporary workaround to unblock updating to Electron 26
* Since you have E2E repro with signal application, I would suggest to investigate the crash attaching a debugger to the renderer process. The partition allocator entry point for render process is at https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/wtf/allocator/partitions.cc, currently there are 3 different partition roots associated with the renderer, we don't know which root is triggering the crash in `partition_alloc::PartitionRoot::PurgeMemory`. So first step would be to identify the root among https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/wtf/allocator/partitions.h;l=116-118 which can give clue to the application code that triggers the crash.

**Comment by indutny-signal** [comment]
I did a small foray into it and found out that Debug build just crashes outside of partition allocator :joy: I can reproduce the crash consistently on Fedora Linux, but if I delete everything from our preload - it doesn't seem to happen.

I suspect that it could be our Neon rust addons that cause the trouble. Will look more into it today!

**Comment by bfurtonmw** [comment]
Adding a datapoint: our users are only running into this crash on shared Linux environments (i.e. multiple users, no root access). I also tested on a Debian 11 VM (VMWare Fusion on iMac host, macOS Ventura), and did not experience the crash. In the past, we've had to implement workarounds on these kind of shared Debian machines due to `setuid` restrictions.

**Comment by threema-danilo** [comment]
> I suspect that it could be our Neon rust addons that cause the trouble. Will look more into it today!

I don't know what the Neon addon does, but we also use a WASM library written in Rust.

Another common thing that both we (Threema) and you (Signal) have is the use of a SQLite fork with a patched-in SQLCipher (not the same fork though).

**Comment by saghul** [comment]
FWIW, we (Jitsi) use neither but our users have experienced the crash too.

**Comment by indutny-signal** [comment]
@deepak1556 I tried backporting the patch and believe that I resolved the conflicts, but the issue is still occuring. Given that it seems that everyone is able to reproduce it, do you need any further action from us?

**Comment by indutny-signal** [comment]
Here's the diff of applied backport FWIW: https://gist.github.com/indutny-signal/707623c1bbbafc065c827c0611fa00d3

**Comment by deepak1556** [comment]
@indutny-signal to confirm, did you add `PartitionAllocMemoryReclaimer` to the `--disable-features` command line flag similar to https://github.com/electron/electron/issues/39775#issuecomment-1714945631 after the backport ?

> Given that it seems that everyone is able to reproduce it, do you need any further action from us?

I tried to repro with the gists from https://github.com/electron/electron/issues/39515 which also share the same stack trace and was not able to confirm the crash in Fedora 36 arm64 vm and Ubuntu 22.04 x86_64 vm. Both vm use 4k page size if that makes any difference. Additionally, there are reports of the crash being addressed with v27 beta, if possible please check if that is the case for your application as well.

@indutny-signal given you are able to crash consistently on fedora, can you share the steps that I can try. Thanks!

**Comment by indutny-signal** [comment]
@deepak1556 I just tried disabling `PartitionAllocBackupRefPtr` with that patched electron version that I built and it still crashes.

I'll try figuring out a minimal repro for it.

**Comment by indutny-signal** [comment]
@deepak1556 reproduces very easily for me with https://gist.github.com/8c61fa47716d73db04ad61fe0201c69e . I just have to start this fiddle and alt-tab a few times until it crashes :)

**Comment by jtbandes** [comment]
I am also able to reproduce the same crash on:
- Ubuntu 22.04.3
- ASUSTeK COMPUTER INC. Zephyrus S15 GX502LXS_GX502LXS
- NV164 / Mesa Intel® UHD Graphics (CML GT2)
- Intel® Core™ i7-10875H CPU @ 2.30GHz × 16
- Node v18.18.0
- Electron 26.2.2

It happens when I use the stock electron-fiddle app, **with the sole addition of `sandbox: false`**. The renderer crashes a few seconds after launching. Stack trace below (I am not sure how to get proper symbols, let me know if anyone knows how):

```
$ ELECTRON_ENABLE_STACK_DUMPING=1 npx electron@26.2.2 .
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
Received signal 11 <unknown> 306d80001048
#0 0x561731a6f263 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x57d7262)
#1 0x561731a7feec (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x57e7eeb)
#2 0x7ff3c1642520 (/usr/lib/x86_64-linux-gnu/libc.so.6+0x4251f)
#3 0x561731b6d7ac (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x58d57ab)
#4 0x561731b67284 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x58cf283)
#5 0x561731a82668 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x57ea667)
#6 0x561730e3a399 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x4ba2398)
#7 0x56173079560a (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x44fd609)
#8 0x56173077b5e9 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x44e35e8)
#9 0x561731a125ed (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x577a5ec)
#10 0x561731a2eb54 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x5796b53)
#11 0x561731a2e369 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x5796368)
#12 0x561731a2f095 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x5797094)
#13 0x5617319cea98 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x5736a97)
#14 0x561731a2f400 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x57973ff)
#15 0x5617319f2dc1 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x575adc0)
#16 0x561735512d54 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x927ad53)
#17 0x56172e8e580b (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x264d80a)
#18 0x56172e8e6cb8 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x264ecb7)
#19 0x56172e8e4527 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x264c526)
#20 0x56172e8e4636 (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x264c635)
#21 0x56172e61821d (/home/foxglove/.npm/_npx/2d0e8517694be14b/node_modules/electron/dist/electron+0x238021c)
#22 0x7ff3c1629d90 (/usr/lib/x86_64-linux-gnu/libc.so.6+0x29d8f)
  r8: 0000000010600002  r9: 80cc4a0020170000 r10: 0000000000000000 r11: 0000000000000001
 r12: 0000000000000000 r13: 00005617361e7008 r14: 0000000000000000 r15: 00005617362990b0
  di: 0000306d80008020  si: 0000000010600002  bp: 00007ffc078713a0  bx: 00005617362990b0
  dx: 0000306d80001040  ax: 0000306d80001040  cx: 0000000000000000  sp: 00007ffc07871390
  ip: 0000561731b68953 efl: 0000000000010246 cgf: 002b000000000033 erf: 0000000000000027
 trp: 000000000000000e msk: 0000000000000000 cr2: 0000306d80001048
[end of stack trace]
[30075:0922/160328.374513:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 1 times!
```


When I use version 27.0.0-beta.1 or newer, it no longer crashes. (I do still see the same `GetVSyncParametersIfAvailable` error, so I guess that's unrelated.)

Let me know if there's any additional information that would be useful for debugging.

I'll cross-post this to #39515 since it seems like a duplicate of this issue.

**Comment by nullie** [comment]
I'm experiencing the same issue on NixOS with Electron 26.2.1 when running Pritunl Client v1.3.3584.5

**Comment by xmedeko** [comment]
@jtbandes AFAIK the error
```
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
```
is different to this one. It also started to appear with v26, but depends on your Linux OS configuration, e.g. see #30902 (older, but people still add comments there.)

**Comment by jtbandes** [comment]
Yes, I think that’s unrelated. It doesn’t prevent launching the app. The app does launch but then the renderer crashes a few seconds later. I just posted the whole log in case there was anything relevant.

**Comment by indutny-signal** [comment]
@deepak1556 is there anything else you'd like me to try out? I'd love to help figure it out!

**Comment by deepak1556** [comment]
@indutny-signal here is my progress so far, I attempted the test case from https://github.com/electron/electron/issues/39775#issuecomment-1726235269 on Ubuntu 22.04 and Fedora 38 but didn't succeed to trigger the crash. However when I switched to use Ubuntu with Sway and started the application in Wayland `--ozone-platform=wayland` I got two crash scenarios for `26-x-y`,

1) Crash on exit inside partition allocator when attempting to free an object that was not allocated by the partition allocator. I couldn't repro this crash with `>=27-x-y` and the crash is fixed when running with `--disable-features=PartitionAllocBackupRefPtr`

<details>
<summary>Stacktrace</summary>

```
#0  0x000055555e6b7fcb in RawLog() () at ../../base/allocator/partition_allocator/partition_alloc_base/logging.cc:278
#1  0x000055555e6b7a01 in  () at ../../base/allocator/partition_allocator/partition_alloc_base/check.cc:86
#2  0x000055555e58167e in FreeNoHooks () at ../../base/allocator/partition_allocator/partition_address_space.h:71
#3  PartitionFree() () at ../../base/allocator/partition_allocator/shim/allocator_shim_default_dispatch_to_partition_alloc.cc:398
#4  0x000055556076a53c in __libcpp_operator_delete<void*> () at ../../buildtools/third_party/libc++/trunk/include/new:278
#5  __do_deallocate_handle_size<> () at ../../buildtools/third_party/libc++/trunk/include/new:302
#6  __libcpp_deallocate () at ../../buildtools/third_party/libc++/trunk/include/new:318
#7  deallocate () at ../../buildtools/third_party/libc++/trunk/include/__memory/allocator.h:130
#8  deallocate () at ../../buildtools/third_party/libc++/trunk/include/__memory/allocator_traits.h:288
#9  ~basic_string () at ../../buildtools/third_party/libc++/trunk/include/string:1066
#10 ~Layer() () at ../../ui/compositor/layer.cc:255
#11 0x000055556076b94e in ui::Layer::~Layer() () at ../../ui/compositor/layer.cc:227
#12 0x000055556079502a in operator() () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:65
#13 reset () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:297
#14 DestroyLayer() () at ../../ui/compositor/layer_owner.cc:95
#15 0x00005555605c2176 in ~Window() () at ../../ui/aura/window.cc:244
#16 0x00005555605c2abe in aura::Window::~Window() () at ../../ui/aura/window.cc:184
#17 0x000055555d7a4f10 in operator() () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:65
#18 reset () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:297
#19 ~WebContentsViewAura() () at ../../content/browser/web_contents/web_contents_view_aura.cc:723
#20 0x000055555d7a511e in content::WebContentsViewAura::~WebContentsViewAura() () at ../../content/browser/web_contents/web_contents_view_aura.cc:715
#21 0x000055555d7682e5 in operator() () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:65
#22 reset () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:297
#23 ~unique_ptr () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:263
#24 ~WebContentsImpl() () at ../../content/browser/web_contents/web_contents_impl.cc:1180
#25 0x000055555d768f0e in content::WebContentsImpl::~WebContentsImpl() () at ../../content/browser/web_contents/web_contents_impl.cc:1069
#26 0x000055555922739d in operator() () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:65
#27 reset () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:297
#28 ~unique_ptr () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:263
#29 ~InspectableWebContents() () at ../../electron/shell/browser/ui/inspectable_web_contents.cc:381
#30 0x000055555922750e in electron::InspectableWebContents::~InspectableWebContents() () at ../../electron/shell/browser/ui/inspectable_web_contents.cc:376
#31 0x0000555559160761 in operator() () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:65
#32 reset () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:297
#33 ~unique_ptr () at ../../buildtools/third_party/libc++/trunk/include/__memory/unique_ptr.h:263
#34 ~WebContents() () at ../../electron/shell/browser/api/electron_api_web_contents.cc:1085
#35 0x0000555559160c0e in electron::api::WebContents::~WebContents() () at ../../electron/shell/browser/api/electron_api_web_contents.cc:1053
#36 0x0000555559160dd1 in DeleteThisIfAlive() () at ../../electron/shell/browser/api/electron_api_web_contents.cc:1097
#37 0x00005555590d323b in Invoke<void (electron::api::BrowserWindow::*)(), base::WeakPtr<electron::api::BrowserWindow> const&> () at ../../base/functional/bind_internal.h:746
#38 MakeItSo<void (electron::api::BrowserWindow::* const&)(), std::__Cr::tuple<base::WeakPtr<electron::api::BrowserWindow> > const&> () at ../../base/functional/bind_internal.h:953
#39 RunImpl<void (electron::api::BrowserWindow::* const&)(), std::__Cr::tuple<base::WeakPtr<electron::api::BrowserWindow> > const&, 0ul> ()
    at ../../base/functional/bind_internal.h:1025
#40 Run() () at ../../base/functional/bind_internal.h:989
#41 0x000055555e4edefe in Run () at ../../base/functional/callback.h:152
#42 RunTaskImpl() () at ../../base/task/common/task_annotator.cc:186
#43 0x000055555e520bd9 in RunTask<(lambda at ../../base/task/sequence_manager/thread_controller_with_message_pump_impl.cc:488:11)> () at ../../base/task/common/task_annotator.h:88
--Type <RET> for more, q to quit, c to continue without paging--
#44 DoWorkImpl() () at ../../base/task/sequence_manager/thread_controller_with_message_pump_impl.cc:486
#45 0x000055555e51fe94 in DoWork() () at ../../base/task/sequence_manager/thread_controller_with_message_pump_impl.cc:351
#46 0x000055555e5218f5 in non-virtual thunk to base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork() ()
#47 0x000055555e5b933b in Run() () at ../../base/message_loop/message_pump_glib.cc:691
#48 0x000055555e521f3b in Run() () at ../../base/task/sequence_manager/thread_controller_with_message_pump_impl.cc:651
#49 0x000055555e4ca21b in Run() () at ../../base/run_loop.cc:134
#50 0x000055555ce8cdbf in RunMainMessageLoop() () at ../../content/browser/browser_main_loop.cc:1067
#51 0x000055555ce8f295 in Run() () at ../../content/browser/browser_main_runner_impl.cc:158
#52 0x000055555ce890ca in BrowserMain() () at ../../content/browser/browser_main.cc:34
#53 0x00005555594520be in RunBrowserProcessMain() () at ../../content/app/content_main_runner_impl.cc:689
#54 0x0000555559454c2a in RunBrowser() () at ../../content/app/content_main_runner_impl.cc:1270
#55 0x00005555594544bf in Run() () at ../../content/app/content_main_runner_impl.cc:1120
#56 0x0000555559450869 in RunContentProcess() () at ../../content/app/content_main.cc:330
#57 0x0000555559450b92 in ContentMain() () at ../../content/app/content_main.cc:347
#58 0x0000555559091bf3 in main () at ../../electron/shell/app/electron_main_linux.cc:40
```
</details>

2) Crash when alt-tab into the application, but this one is same as https://github.com/electron/electron/issues/39449

 Although crash cases reported here and in https://github.com/electron/electron/issues/39515 are when `sandbox: false` for the renderer and not the main process. I saw the bisect ranges attached in https://github.com/electron/electron/issues/39515#issuecomment-1688830738 but the only thing that stood out was `BackupRefPtr` feature enabled on Linux and again this is not enabled for renderer process so should be unrelated.

@indutny-signal can you attach the crash stack for https://github.com/electron/electron/issues/39775#issuecomment-1726235269 ? I will add some crash keys to the relevant call sites in a custom electron build as next step.

**PS:** If anyone can provide a reliable repro case using a clean VM even without electron fiddle but rather using a production application, I am happy to set it up.

**Comment by deepak1556** [comment]
I am attaching a dcheck build of Electron v26.2.2 for linux x64 https://github.com/deepak1556/electron-debug-version/releases/tag/v26.2.2, for those who are able to repro the issue please run this build with your repro fiddle with `--enable-logging` and attach the generated logs, crash dumps to this issue thread.

**Comment by threema-danilo** [comment]
@deepak1556 thanks for the build! I first tested whether I can still reproduce the issue with 26.2.2. I put "26.2.2" in `package.json`, started [the app](https://threema.ch/en/download-md) and immediately got a renderer crash.

Then I unpacked your build:

```
$ cd node_modules/electron/dist
$ rm -r *
$ unzip /tmp/electron-v26.2.2-linux-x64-dcheck.zip
```

...and rebuilt our app with it. Now the app takes extremely long to start (I assume due to all the debug logic that was enabled in this build), and... doesn't crash anymore 🙃

As a sanity check, I ran `rm -rf node_modules/ && npm install`, started the app again, and it crashes again.

The parameters I used for starting: `--enable-logging=stderr`.

---

When the application starts, usually a lot of "ui <-> worker" and "worker <-> database" communication takes place. Maybe the debug logic slowed down the app, which resulted in a different memory allocation behavior, and no crash. Or maybe it's a race condition, that wasn't triggered because the code was slower.

Is the log still helpful if the crash wasn't reproduced? I assume not?

**Comment by threema-danilo** [comment]
It seems that I cannot reproduce the crash with `27.0.0-beta.6`.

**Comment by deepak1556** [comment]
@threema-danilo thanks for testing, yeah the logs from the run will not be useful without the crash being triggered. I am hoping users who reported crash with `electron-fiddle` are able to trigger it with the debug builds.

@xmedeko yes based on the reports so far, the issue does not happen with v27.

**Comment by deepak1556** [comment]
Great, can you attach the coredump file for the above fiddle test. Also, can you check if the [custom build](https://github.com/deepak1556/electron-debug-version/releases/tag/v26.2.2) is able to trigger the crash for the same fiddle test case, if so attach its coredump file as well.

**Comment by threema-danilo** [comment]
@deepak1556 ah, I just found out how to use custom Electron builds in Fiddle. I'm happy to test it! However: How do I get a crashpad dump in Fiddle? I tried the following:

```ts
crashReporter.start({
  submitURL: 'https://example.invalid/crash-report',
  uploadToServer: false,
});

...

app.whenReady().then(() => {
  app.setPath('crashDumps', '/tmp/crashes');
});
```

...but no crash dumps are generated.

Or should I simply attach the regular sytemd coredump instead?

**Comment by deepak1556** [comment]
You can add the following to the fiddle before `app.whenReady` and minidump should be available at the path specified. If you want to change paths, update it before `crashReporter.start`

```
console.log(app.getPath('crashDumps'))
crashReporter.start({ submitURL: '', uploadToServer: false })
```

> should I simply attach the regular sytemd coredump instead?

Yes you can also attach this.

**Comment by threema-danilo** [comment]
> Related, are you able to repro the issue using https://gist.github.com/8c61fa47716d73db04ad61fe0201c69e as fiddle test case with normal v26.2.2 builds ?

Huh, I cannot reproduce the issue with the dcheck build. It reliably crashes with the regular 26.2.2 though.

Here's the minidump for 26.2.2: [62b76819-9ebe-4f4c-9d33-b9486dc5f45e.dmp.zip](https://github.com/electron/electron/files/12746972/62b76819-9ebe-4f4c-9d33-b9486dc5f45e.dmp.zip)

Symbolicated backtrace:

```
Thread 0 electron (crashed)
 0  electron!partition_alloc::internal::PartitionBucket::MaintainActiveList() [partition_bucket.cc : 1163 + 0x0]
     rax = 0x00002b5ec0001040    rdx = 0x00002b5ec0001040
     rcx = 0x0000000000000000    rbx = 0x000055b53b9720b0
     rsi = 0x0000000010600002    rdi = 0x00002b5ec0008020
     rbp = 0x00007ffec7b6af90    rsp = 0x00007ffec7b6af80
      r8 = 0x0000000010600002     r9 = 0x80cc4a00640f0000
     r10 = 0x0000000000000000    r11 = 0x0000000000000001
     r12 = 0x0000000000000000    r13 = 0x000055b53b8c0008
     r14 = 0x0000000000000000    r15 = 0x000055b53b9720b0
     rip = 0x000055b537241953
    Found by: given as instruction pointer in context
 1  electron!partition_alloc::PartitionRoot::PurgeMemory(int) [partition_root.cc : 1380 + 0x7]
     rbx = 0x0000000000000003    rbp = 0x00007ffec7b6b420
     rsp = 0x00007ffec7b6afa0    r12 = 0x0000000000000000
     r13 = 0x000055b53b8c0008    r14 = 0x000055b53b973488
     r15 = 0x000055b53b9720b0    rip = 0x000055b5372467ac
    Found by: call frame info
 2  electron!partition_alloc::MemoryReclaimer::Reclaim(int) [memory_reclaimer.cc : 89 + 0x7]
     rbx = 0x000055b53b8c0008    rbp = 0x00007ffec7b6b450
     rsp = 0x00007ffec7b6b430    r12 = 0x00000f64002c3420
     r13 = 0x000055b53b8c0008    r14 = 0x0000000000000003
     r15 = 0x000055b53b8c0018    rip = 0x000055b537240284
    Found by: call frame info
 3  electron!base::allocator::(anonymous namespace)::RunMemoryReclaimer(scoped_refptr<base::SequencedTaskRunner>) [partition_alloc_support.cc : 227 + 0x7]
     rbx = 0x00000f6400239730    rbp = 0x00007ffec7b6b4c0
     rsp = 0x00007ffec7b6b460    r12 = 0x00000000c1f115d8
     r13 = 0x000055b53b8c0008    r14 = 0xaaaaaaaaaaaaaaaa
     r15 = 0x00007ffec7b6b500    rip = 0x000055b53715b668
    Found by: call frame info
...
```

**Comment by deepak1556** [comment]
@threema-danilo can you test with the following [debug build](https://output.circle-artifacts.com/output/job/b5bf1351-f842-487e-b93c-8a9ee813de1b/artifacts/0/linux-x64/dist.zip) as well, this is from our CI with latest 26-x-y. The build I created had [enable_pkeys](https://source.chromium.org/chromium/chromium/src/+/main:base/allocator/partition_allocator/partition_alloc.gni;l=318-324) disabled which avoids creating a partition allocator pool used for V8 CFI. I had it disabled to investigate another crash but now I am curious if that also had impact on this bug or is it just debug builds generally avoid the crash.

**Comment by deepak1556** [comment]
Friendly ping, if anyone else is able to reliably repro the issue please test the fiddle https://gist.github.com/8c61fa47716d73db04ad61fe0201c69e with following two debug builds and post the results. It would help move this issue forward, thanks!

[Debug build with threaded allocator disabled](https://github.com/deepak1556/electron-debug-version/releases/tag/v26.2.2)
[Debug build](https://output.circle-artifacts.com/output/job/b5bf1351-f842-487e-b93c-8a9ee813de1b/artifacts/0/linux-x64/dist.zip)

**Comment by indutny-signal** [comment]
Sorry for the delay. I'll give it a go tomorrow!

**Comment by indutny-signal** [comment]
@deepak1556 the debug build with threaded allocator disabled doesn't crash, and the second debug build link has expired or otherwise unavailable!

**Comment by deepak1556** [comment]
Sorry about the expired link, I have now republished the CI build [here](https://github.com/deepak1556/electron-debug-version/releases/tag/v26.2.4)

**Comment by indutny-signal** [comment]
@deepak1556 sorry, I only got around to it no. The build in your last comment crashes for me. Here's the stack trace: https://gist.github.com/indutny-signal/7bfdbc44dfd8fdad9513facee33cb5f5

**Comment by threema-danilo** [comment]
> @threema-danilo can you test with the following [debug build](https://output.circle-artifacts.com/output/job/b5bf1351-f842-487e-b93c-8a9ee813de1b/artifacts/0/linux-x64/dist.zip) as well, this is from our CI with latest 26-x-y.

Sorry, I was travelling for 2 weeks.

I just tested with the fiddle at https://gist.github.com/8c61fa47716d73db04ad61fe0201c69e. Just like @indutny-signal reported:

- [Debug build with threaded allocator disabled](https://github.com/deepak1556/electron-debug-version/releases/tag/v26.2.2): Does not crash
- [DCHECK build for v26.2.4](https://github.com/deepak1556/electron-debug-version/releases/tag/v26.2.4): Crashes with this [stack trace](https://gist.github.com/threema-danilo/c7c6d05b24de74377f98e5895ac4d31c)

**Comment by andrewheadricke** [comment]
No longer an issue with the recently released 27.0.0

**Comment by deepak1556** [comment]
Thanks @indutny-signal and @threema-danilo for the test results, they are helpful. I have pinged Chromium channel with the data gathered so far. I will update here once there are some next steps.

**Comment by deepak1556** [comment]
Addressed in https://github.com/electron/electron/pull/40377, fix available with >= 26.4.3
